dll/win32/shell32/dialogs/item_prop.cpp

```diff
@@ -0,0 +1,182 @@
+        {
+            FreeData(pData);
+            return E_OUTOFMEMORY;
+        }
+        if (pClsidDefault)
+        {
+            pData->ClsidDefault = *pClsidDefault;
+            pData->pClsidDefault = &pData->ClsidDefault;
+        }
+
+        HRESULT hr = S_OK;
+        if (pDO)
+            hr = CoMarshalInterThreadInterfaceInStream(IID_IDataObject, pDO, &pData->pObjStream);
+
+        UINT flags = CTF_COINIT | CTF_PROCESS_REF | CTF_INSIST;
+        if (SUCCEEDED(hr) && !SHCreateThread(ShowPropertiesThread, pData, flags, NULL))
+        {
+            if (pData->pObjStream)
+                pData->pObjStream->Release();
+            hr = E_FAIL;
+        }
+        if (FAILED(hr))
+            FreeData(pData);
+        return hr;
+    }
+
+    static DWORD CALLBACK ShowPropertiesThread(LPVOID Param)
+    {
+        DATA *pData = (DATA*) Param;
+        CComPtr<IDataObject> pDO;
+        if (pData->pObjStream)
+            CoGetInterfaceAndReleaseStream(pData->pObjStream, IID_PPV_ARG(IDataObject, &pDO));
+        Show(pData->pClsidDefault, pDO, pData->InitFunc, pData->InitString);
+        FreeData(pData);
+        return 0;
+    }
+};
+
+static void CALLBACK
+FSFolderItemPropDialogInitCallback(LPCWSTR InitString, IDataObject *pDO, HKEY *hKeys, UINT *cKeys)
+{
+    // Add file-type specific pages
+    UNREFERENCED_PARAMETER(InitString);
+    CDataObjectHIDA cida(pDO);
+    if (SUCCEEDED(cida.hr()) && cida->cidl)
+    {
+        PCUITEMID_CHILD pidl = HIDA_GetPIDLItem(cida, 0);
+        AddFSClassKeysToArray(1, &pidl, hKeys, cKeys);
+    }
+}
+
+static void CALLBACK
+ClassPropDialogInitCallback(LPCWSTR InitString, IDataObject *pDO, HKEY *hKeys, UINT *cKeys)
+{
+    // Add pages from HKCR\%ProgId% (with shellex\PropertySheetHandlers appended later)
+    UNREFERENCED_PARAMETER(pDO);
+    AddClassKeyToArray(InitString, hKeys, cKeys);
+}
+
+HRESULT
+SHELL32_ShowFilesystemItemPropertiesDialogAsync(IDataObject *pDO)
+{
+    CDataObjectHIDA cida(pDO);
+    HRESULT hr = cida.hr();
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    const CLSID *pClsid = NULL;
+    ShellPropSheetDialog::PFNINITIALIZE InitFunc = NULL;
+    LPCWSTR InitString = NULL;
+
+    if (_ILIsDrive(HIDA_GetPIDLItem(cida, 0)))
+    {
+        pClsid = &CLSID_ShellDrvDefExt;
+        InitFunc = ClassPropDialogInitCallback;
+        InitString = L"Drive";
+    }
+    else
+    {
+        pClsid = &CLSID_ShellFileDefExt;
+        InitFunc = FSFolderItemPropDialogInitCallback;
+    }
+    ShellPropSheetDialog Dialog;
+    return Dialog.ShowAsync(pClsid, pDO, InitFunc, InitString);
+}
+
+HRESULT
+SHELL32_ShowShellExtensionProperties(const CLSID *pClsid, IDataObject *pDO)
+{
+    WCHAR ClassBuf[6 + 38 + 1] = L"CLSID\\";
+    StringFromGUID2(*pClsid, ClassBuf + 6, 38 + 1);
+    ShellPropSheetDialog Dialog;
+    return Dialog.ShowAsync(NULL, pDO, ClassPropDialogInitCallback, ClassBuf);
+}
+
+HRESULT
+SHELL_ShowItemIDListProperties(LPCITEMIDLIST pidl)
+{
+    if (!pidl)
```

Comment:
Why not just asserting on pidl being (not) NULL ?
I would expect here that the callers of this SHELL_ShowItemIDListProperties helper don't do weird things?
(it seems to be `CFSFolder::CallBack` unconditionally calling `SHELL_ShowItemIDListProperties(m_pidlRoot);` with a `m_pidlRoot` that may be NULL; but if so, is it correct?)

Reply:
`CFSFolder::Initialize` has rather poor error checking but that is no excuse, `m_pidlRoot` is assumed to not be NULL in this specific case.

CFSFolder is not meant to be the only caller (the Recycle Bin background context menu will be another later) but if you are happy with relying on callers not doing something stupid, I'll change this to an assert.

Reply:
As long as these callers are internal shell32 functions, yes I would expect they don't do the stupid.